### PR TITLE
[SMTChecker] Implement Boolean short-circuiting correctly

### LINKED ISSUE #3711: SMT: Check that short-circuiting is implemented correctly.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #3711

Depends on #6371

### CODE REVIEW

File: libsolidity/formal/SMTChecker.cpp [hunk 1]
```diff
@@ -1456,7 +1473,12 @@ TypePointer SMTChecker::typeWithoutPointer(TypePointer const& _type)
 void SMTChecker::mergeVariables(vector<VariableDeclaration const*> const& _variables, smt::Expression const& _condition, VariableIndices const& _indicesEndTrue, VariableIndices const& _indicesEndFalse)
 {
 	set<VariableDeclaration const*> uniqueVars(_variables.begin(), _variables.end());
-	for (auto const* decl: uniqueVars)
+	mergeVariables(uniqueVars, _condition, _indicesEndTrue, _indicesEndFalse);
+}
+
+void SMTChecker::mergeVariables(set<VariableDeclaration const*> const& _variables, smt::Expression const& _condition, VariableIndices const& _indicesEndTrue, VariableIndices const& _indicesEndFalse)
+{
+	for (auto const* decl: _variables)
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp:1481]: Can that lead to nondeterminism?

**Comment by chriseth** [comment]
Looks good!

File: libsolidity/formal/SMTChecker.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp:1481]: What exactly, the order of traversal? Don't think so, their SSA index doesn't depend on that.

Comment (chriseth) [reply libsolidity/formal/SMTChecker.cpp:1481]: What about the "SMT aux string"?

Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp:1481]: Tracked in #6411

**Comment by leonardoalt** [comment]
Rebased

**Pull request merged**